plugins/rtmp-services/dacast.c

```diff
@@ -0,0 +1,237 @@
+
+	if (!item_video_codec || !item_h264_profile || !item_scanning ||
+	    !item_rate_control || !item_audio_codec || !item_frame_rate ||
+	    !item_keyframe_interval || !item_audio_bitrate ||
+	    !item_audio_channels || !item_audio_sample_rate)
+		goto finish;
+
+	const char *video_codec = json_string_value(item_video_codec);
+	const char *h264_profile = json_string_value(item_h264_profile);
+	const char *scanning = json_string_value(item_scanning);
+	const char *rate_control = json_string_value(item_rate_control);
+	const char *audio_codec = json_string_value(item_audio_codec);
+
+	double frame_rate = json_number_value(item_frame_rate);
+	int keyframe_interval = json_integer_value(item_keyframe_interval);
+	int audio_bitrate = json_integer_value(item_audio_bitrate);
+	int audio_channels = json_integer_value(item_audio_channels);
+	int audio_sample_rate = json_integer_value(item_audio_sample_rate);
+
+	if (strcmp(video_codec, "H.264") == 0) {
+		info->ingest.video_codec = bstrdup("x264");
+	}
+	info->ingest.h264_profile = bstrdup(h264_profile);
+	info->ingest.scanning = bstrdup(scanning);
+	info->ingest.rate_control = bstrdup(rate_control);
+	info->ingest.audio_codec = bstrdup(audio_codec);
+
+	info->ingest.frame_rate = frame_rate;
+	info->ingest.keyframe_interval = keyframe_interval / 60;
+	info->ingest.audio_bitrate = audio_bitrate;
+	info->ingest.audio_channels = audio_channels;
+	info->ingest.audio_sample_rate = audio_sample_rate;
+
+	info->last_time = os_gettime_ns() / SEC_TO_NSEC;
+
+	success = true;
+
+finish:
+	if (root)
+		json_decref(root);
+	return success;
+}
+
+static bool dacast_ingest_update(void *param, struct file_download_data *data)
+{
+	bool success;
+
+	pthread_mutex_lock(&mutex);
+	success = load_ingests((const char *)data->buffer.array,
+			       (const char *)param);
+	pthread_mutex_unlock(&mutex);
+
+	update_info_destroy(dacast_update_info);
+	dacast_update_info = NULL;
+
+	if (success) {
+		os_atomic_set_bool(&ingests_loaded, true);
+	}
+
+	return true;
+}
+
+struct dacast_ingest *dacast_ingest(const char *key)
+{
+	pthread_mutex_lock(&mutex);
+	struct dacast_ingest_info *info = find_ingest(key);
+	pthread_mutex_unlock(&mutex);
+	return info == NULL ? &dacast_invalid_ingest : &info->ingest;
+}
+
+void init_dacast_data(void)
+{
+	da_init(cur_ingests);
+	pthread_mutex_init(&mutex, NULL);
+}
+
+extern const char *get_module_name(void);
+
+void dacast_ingests_load_data(const char *server, const char *key)
+{
+	struct dstr uri = {0};
+
+	os_atomic_set_bool(&ingests_loaded, false);
+
+	dstr_copy(&uri, server);
+	dstr_cat(&uri, key);
+
+	dacast_update_info = update_info_create_single(
+		"[dacast ingest load data] ", get_module_name(), uri.array,
+		dacast_ingest_update, (void *)key);
+
+	if (!os_atomic_load_bool(&ingests_loaded)) {
+		for (int i = 0; i < 3 * 100; i++) {
+			if (os_atomic_load_bool(&ingests_loaded)) {
+				break;
+			}
+			os_sleep_ms(10);
+		}
+	}
+
+	dstr_free(&uri);
+}
```

Comment:
What happens if this API request fails?

Reply:
I believe in this case it _should_ trigger the same error when you click **Start Streaming** as it would for an invalid key, i.e.

![image](https://user-images.githubusercontent.com/1795615/103531644-b0452000-4e81-11eb-8e47-a8f6b727f015.png)

If this is acceptable please let me know so we can make this change.